### ISSUE TITLE
Issue: Forward pass of mamba block results in 0s on a Quadro P6000 GPU

### ISSUE THREAD

**Issue by ICharlotteI** [state-spaces/mamba#225]
Once I had my mamba environment set up and mamba-ssm installed successfully, I ran the example code provided by the repository and the output of x and y was as follows: 
tensor([[[-1.9903e+00, -9.2922e-01, -2.8890e+00,  ..., -2.4487e-03,
          -1.4836e+00, -1.7656e+00],
         [-6.2470e-01, -4.3941e-01,  1.2152e+00,  ..., -1.2207e-01,
          -2.0163e-02, -2.1134e+00],
         [-1.2493e+00, -4.5132e-01,  6.3297e-01,  ..., -1.6900e+00,
          -1.4243e+00, -4.9173e-01],
         ...,
         [ 1.1544e-01, -1.4427e+00, -3.0226e-01,  ...,  7.3694e-01,
          -3.7878e-01,  2.1897e+00],
         [ 6.6608e-01, -2.5774e-01, -9.2452e-01,  ..., -2.6071e-02,
           1.2988e+00, -1.8465e+00],
         [-7.5659e-01, -1.3786e+00, -8.3231e-01,  ...,  9.0802e-01,
          -1.7450e-01, -1.8339e+00]],

        [[ 6.3811e-01, -2.0693e+00, -8.9093e-01,  ...,  1.8312e-01,
           1.1453e-01,  4.1747e-01],
         [-1.1871e+00,  2.0305e-01,  5.3381e-01,  ..., -2.3646e-01,
           1.2816e+00, -5.6714e-01],
         [-1.6325e+00, -1.0873e+00, -4.7420e-01,  ...,  1.3370e+00,
           2.8081e-02, -9.4508e-02],
         ...,
         [-3.5926e-01, -4.8741e-01, -1.1013e+00,  ..., -2.1140e-01,
          -3.3222e-01,  3.1988e-01],
         [-1.0671e+00, -3.9777e-02,  1.3775e+00,  ...,  4.9738e-01,
           4.1687e-01, -4.6370e-01],
         [-6.4432e-01, -1.6179e+00, -2.2629e-01,  ...,  8.0204e-01,
           1.1772e+00, -3.3466e-02]]], device='cuda:0') 
tensor([[[0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         ...,
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.]],

        [[0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         ...,
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.],
         [0., 0., 0.,  ..., 0., 0., 0.]]], device='cuda:0',
       grad_fn=<MambaInnerFnBackward>)
 Is there anyone who has encountered the same problem with the output of 0? My GPU is a Quadro P6000.

**Comment by vasqu** [comment]
The 0s are only the backward pass which I assume you haven't done, e.g. using a loss with the output and calling .backward on the loss. See the following example to construct a (nonsensical) dummy loss:
```python
import torch
from mamba_ssm import Mamba

batch, length, dim = 2, 64, 16

# assuming you're using the mamba block?
model = Mamba(
    # This module uses roughly 3 * expand * d_model^2 parameters
    d_model=dim, # Model dimension d_model
    d_state=16,  # SSM state expansion factor
    d_conv=4,    # Local convolution width
    expand=2,    # Block expansion factor
).to("cuda")
optimizer = torch.optim.SGD(model.parameters(), lr=0.01, momentum=0.9)

model.zero_grad()
optimizer.zero_grad()

x = torch.randn(batch, length, dim).to("cuda")
y = model(x)
assert y.shape == x.shape

# some nonsensical loss --> the part which you didn't do, hence a backward of 0 is to be expected
loss = torch.log(1 + torch.abs(y.sum()))
loss.backward()

# print your values along the gradients / backward values ... , for example print(y)
```

If you meant something else, could you provide the complete code snippet of yours. I'm not sure if you really only tested the mamba block or something else along the way.

**Comment by ICharlotteI** [comment]
> The 0s are only the backward pass which I assume you haven't done, e.g. using a loss with the output and calling .backward on the loss. See the following example to construct a (nonsensical) dummy loss:
> 
> ```python
> import torch
> from mamba_ssm import Mamba
> 
> batch, length, dim = 2, 64, 16
> 
> # assuming you're using the mamba block?
> model = Mamba(
>     # This module uses roughly 3 * expand * d_model^2 parameters
>     d_model=dim, # Model dimension d_model
>     d_state=16,  # SSM state expansion factor
>     d_conv=4,    # Local convolution width
>     expand=2,    # Block expansion factor
> ).to("cuda")
> optimizer = torch.optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
> 
> model.zero_grad()
> optimizer.zero_grad()
> 
> x = torch.randn(batch, length, dim).to("cuda")
> y = model(x)
> assert y.shape == x.shape
> 
> # some nonsensical loss --> the part which you didn't do, hence a backward of 0 is to be expected
> loss = torch.log(1 + torch.abs(y.sum()))
> loss.backward()
> 
> # print your values along the gradients / backward values ... , for example print(y)
> ```
> 
> If you meant something else, could you provide the complete code snippet of yours. I'm not sure if you really only tested the mamba block or something else along the way.

Thank you. My test code is:
import torch
from mamba_ssm import Mamba

batch, length, dim = 2, 64, 16
x = torch.randn(batch, length, dim).to("cuda")
model = Mamba(
    # This module uses roughly 3 * expand * d_model^2 parameters
    d_model=dim, # Model dimension d_model
    d_state=16,  # SSM state expansion factor
    d_conv=4,    # Local convolution width
    expand=2,    # Block expansion factor
).to("cuda")
y = model(x)
assert y.shape == x.shape
print(x,y)

But I think it's not a problem of not having a reverse update, because I'm printing the model output, not the gradient, so normally it should have a value. Also, I used the same mamba and causal convolutional versions on different machines and got different results. For example, here's what I printed on a 2080Ti:
tensor([[[ 1.1041,  0.5550, -0.2541,  ...,  0.4169,  1.1083,  1.0599],
         [ 0.0142, -0.8575, -0.0492,  ...,  1.8348, -1.4518,  0.8332],
         [ 0.3599, -1.1643,  0.2354,  ..., -0.4126, -1.7095, -1.2917],
         ...,
         [-0.1527,  0.5961, -0.4259,  ...,  0.1916, -0.9371, -0.9781],
         [ 0.0612,  0.8239, -1.4445,  ..., -1.1242,  0.7850, -1.2839],
         [ 0.3485, -0.1858, -1.1279,  ...,  0.1867, -0.3919, -0.1363]],

        [[ 2.8846, -0.7248, -0.1769,  ...,  0.3879, -0.3978,  2.8736],
         [ 0.9959, -1.0812,  0.6109,  ...,  0.6347,  0.6751,  0.0801],
         [ 0.4403, -1.2142,  0.7594,  ..., -0.7863,  1.2178,  0.1174],
         ...,
         [-0.2930,  0.0820,  0.8295,  ...,  0.9911,  0.9432,  0.2254],
         [ 1.6817,  0.4239,  1.6997,  ...,  1.1775,  0.9638, -0.3200],
         [-0.2192, -0.8363, -0.4727,  ..., -0.4999, -0.9731,  0.0757]]],
       device='cuda:0') 
tensor([[[-0.0317,  0.0441, -0.0098,  ..., -0.0156, -0.0007,  0.0070],
         [ 0.0436,  0.0350, -0.0510,  ...,  0.0328, -0.0013,  0.0364],
         [-0.0110,  0.0344, -0.0262,  ...,  0.0277,  0.0066,  0.0047],
         ...,
         [-0.0305,  0.0216, -0.0011,  ..., -0.0222, -0.0199, -0.0103],
         [-0.0166,  0.0071,  0.1274,  ..., -0.0112,  0.0423,  0.0281],
         [-0.0011,  0.0247, -0.0023,  ..., -0.0097,  0.0560, -0.0150]],

        [[-0.0403, -0.0057, -0.0281,  ...,  0.0160, -0.0355,  0.0047],
         [-0.0451, -0.0700, -0.0343,  ..., -0.0112, -0.0294,  0.0233],
         [ 0.0226,  0.0164, -0.0280,  ...,  0.0017,  0.0624,  0.0727],
         ...,
         [ 0.0278,  0.0207, -0.0571,  ..., -0.0091,  0.0157,  0.0481],
         [-0.0511, -0.0385,  0.0117,  ...,  0.0043, -0.0456, -0.0183],
         [-0.0577, -0.0306, -0.0087,  ..., -0.0112, -0.0712,  0.0870]]],
       device='cuda:0', grad_fn=<MambaInnerFnBackward>)
       
It's worth noting that I'm using the same pytorch version (1.13.1), cuda version (11.7), mamba version (1.1.13.post1), etc.

**Comment by vasqu** [comment]
Hey @ICharlotteI, thanks for the additional information and even testing it on a different gpu. Tbh, I'm also quite confused as to what's going on then. You're completely right that it doesn't have anything to do with the backward pass. Hopefully, someone else with more experience with cuda can help you there but just some thoughts;
- Is it reproducible on your machine? (with the quadro gpu)
- Have you tried different versions of `causal-conv1d` (1.1) and `mamba-ssm` (1.1.12) for example.
- Can you debug the forward pass, i.e. look into what step might cause the output to be zero.
- Maybe change the title of the issue to something like `Forward pass of mamba block results in 0s on a Quadro P6000 GPU` --> just a suggestion to amplify the GPU issue.

**Comment by ICharlotteI** [comment]
> Hey @ICharlotteI, thanks for the additional information and even testing it on a different gpu. Tbh, I'm also quite confused as to what's going on then. You're completely right that it doesn't have anything to do with the backward pass. Hopefully, someone else with more experience with cuda can help you there but just some thoughts;
> 
> * Is it reproducible on your machine? (with the quadro gpu)
> * Have you tried different versions of `causal-conv1d` (1.1) and `mamba-ssm` (1.1.12) for example.
> * Can you debug the forward pass, i.e. look into what step might cause the output to be zero.
> * Maybe change the title of the issue to something like `Forward pass of mamba block results in 0s on a Quadro P6000 GPU` --> just a suggestion to amplify the GPU issue.

Thank you for your advice.
In fact, I have now reproduced MambaBlock and implemented a simple prediction task to evaluate the performance. The results on the P6000 are similar to those based on Mamba on the 2080Ti, but the memory usage is much larger than the official version.
Yes, I have tried other mamba-ssm and causal-conv1d versions. Initially I used the latest version of 1.2.0.post1, but the same result still occurs. But I have not tried the version you mentioned. I will try it later.
As for the forward pass, I am currently just simply debugging the forward pass process in mamba_simple.py, and initially found that the problem appears in:
            if causal_conv1d_fn is None:
                x = self.act(self.conv1d(x)[..., :seqlen])
            else:
                assert self.activation in ["silu", "swish"]
                x = causal_conv1d_fn(
                    x=x,
                    weight=rearrange(self.conv1d.weight, "d 1 w -> d w"),
                    bias=self.conv1d.bias,
                    activation=self.activation,
                )
and 
           out = mamba_inner_fn(
                xz,
                self.conv1d.weight,
                self.conv1d.bias,
                self.x_proj.weight,
                self.dt_proj.weight,
                self.out_proj.weight,
                self.out_proj.bias,
                A,
                None,  # input-dependent B
                None,  # input-dependent C
                self.D.float(),
                delta_bias=self.dt_proj.bias.float(),
                delta_softplus=True,
            )
In both codes, the output is 0. And even if I uninstall causal_conv1d, the output is still 0.
Thank you again for your suggestion on the title, I will change the title, but I don't know if it is the problem of GPU, or the problem of the packet of causal convolution.

**Comment by vasqu** [comment]
Depending on how you reimplemented the mamba block I'd expect less efficiency (memory and speed wise) as you won't have all cuda optimizations.

As you find the problem to be around the `causal_conv1d_fn` and the `mamba_inner_fn` function, I'd suspect the causal conv1d to be the cause. I've seen you've also tried it without the conv1d package but could you do that additionally after adding to [L72](https://github.com/state-spaces/mamba/blob/34076d664838588a3c97727b263478ab9f621a07/mamba_ssm/modules/mamba_simple.py#L72) a `nn.init` (e.g. `nn.init.unfiform_(self.conv1d), a=2, b=5`). This is to see, if the weights have any impact on the outcome and if there maybe there is an inherent issue in torch as well. Furthermore, you could isolate the suspected functions and try them under different conditions (i.e. randomly generate inputs for the funtions and control how the output reacts).

**Comment by vasqu** [comment]
If indeed `nn.conv1d` and the cuda `causal_conv1d` turn out to be the issue, I would even suggest submitting an issue in the torch repo + causal_conv1d repo.

**Comment by ICharlotteI** [comment]
> If indeed `nn.conv1d` and the cuda `causal_conv1d` turn out to be the issue, I would even suggest submitting an issue in the torch repo + causal_conv1d repo.

Thank you for your reply. I will debug further and try out the suggestions you give. If I find a solution to the problem, I will reply under this issue.

**Comment by LIUQI-creat** [comment]
Hello, did you solve the problem, I'm having the same problem as you. I have pytorch version (1.13.0),cuda version (11.7) and GPU is P40.
But I found a mind-boggling thing during the step-by-step debugging process: **adding a `print(z)` above line 189 in mamba_simple.py** gives me non-zero results in the forward pass.
But z as well as the model output will immediately become **nan**

**Comment by changjh23** [comment]
Hi, did you solve this problem? I have the same GPU as you and encountered an incompatibility issue with arithmetic (sm60). May I ask what version of causl_conv1d and mamba_ssm you have installed?

**Comment by colorfulfuture** [comment]
> Hi, did you solve this problem? I have the same GPU as you and encountered an incompatibility issue with arithmetic (sm60). May I ask what version of causl_conv1d and mamba_ssm you have installed?

Hi,

I think the reason is about the GPU Architecture because I tried different GPUs (V100, A100, P100, K4) in my university cluster with the same CUDA version, but some of them failed and output zeros,

I think it is the same problem of flash-attention which can only be used in A series GPU.

**Comment by changjh23** [comment]
> > Hi, did you solve this problem? I have the same GPU as you and encountered an incompatibility issue with arithmetic (sm60). May I ask what version of causl_conv1d and mamba_ssm you have installed?
> 
> Hi,
> 
> I think the reason is about the GPU Architecture because I tried different GPUs (V100, A100, P100, K4) in my university cluster with the same CUDA version, but some of them failed and output zeros,
> 
> I think it is the same problem of flash-attention which can only be used in A series GPU.

oh ,ok. I'm going to try the new GPU as well, thank you very much for your reply.

**Comment by sandhyat** [comment]
> > Hi, did you solve this problem? I have the same GPU as you and encountered an incompatibility issue with arithmetic (sm60). May I ask what version of causl_conv1d and mamba_ssm you have installed?
> 
> Hi,
> 
> I think the reason is about the GPU Architecture because I tried different GPUs (V100, A100, P100, K4) in my university cluster with the same CUDA version, but some of them failed and output zeros,
> 
> I think it is the same problem of flash-attention which can only be used in A series GPU.

Hello @colorfulfuture, 

Could you please share if you faced this issue on V100s too or if the code worked on them? 

Thanks,
Sandhya

**Comment by ICharlotteI** [comment]
I agree with the idea that the cause is related to the GPU architecture, but since it's not experimentally verified there's no guarantee that it's correct. I'm currently running Mamba on a 4090 GPU and everything works fine.
